File: agent/tsdb/src/FeatBit.TsdbServer/Models/Dtos.cs
using System.Text.Json.Serialization;

namespace FeatBit.TsdbServer.Models;

/// <summary>
/// Top-level payload sent by FeatBit SDK — one or more insight batches.
/// Identical to the TrackPayload in agent/data (shared SDK contract).
/// </summary>
public sealed class TrackPayload
{
    public EndUserDto User { get; set; } = default!;
    public IReadOnlyList<FlagEvalDto> Variations { get; set; } = [];
    public IReadOnlyList<MetricEventDto> Metrics { get; set; } = [];

    public bool IsValid() => User is { KeyId.Length: > 0 };
}

public sealed class EndUserDto
{
    public string KeyId { get; set; } = string.Empty;
    public string? Name { get; set; }
    public Dictionary<string, string>? Properties { get; set; }
}

public sealed class FlagEvalDto
{
    public string FlagKey { get; set; } = string.Empty;
    public string Variant { get; set; } = string.Empty;
    public bool SendToExperiment { get; set; }
    public string? ExperimentId { get; set; }
    public string? LayerId { get; set; }
    public long Timestamp { get; set; } // unix seconds
}

public sealed class MetricEventDto
{
    public string EventName { get; set; } = string.Empty;
    public double? NumericValue { get; set; }
    public string? Type { get; set; }
    public string? Route { get; set; }
    public string? AppType { get; set; }
    public long Timestamp { get; set; } // unix seconds
    public Dictionary<string, object>? Props { get; set; }
}

// ── Query request / response DTOs ─────────────────────────────────────────────

/// <summary>
/// HTTP request body for POST /api/query/experiment.
/// Mirrors ExperimentParams in agent/data for backward compatibility.
/// </summary>
public sealed class ExperimentQueryRequest
{
    public required string EnvId { get; init; }
    public required string FlagKey { get; init; }
    public required string MetricEvent { get; init; }
    public required string MetricType { get; init; }    // binary | revenue | count | duration
    public string MetricAgg { get; init; } = "once";     // once | sum | mean | count | latest
    public required string ControlVariant { get; init; }
    public required string TreatmentVariant { get; init; }
    public required DateTimeOffset Start { get; init; }
    public required DateTimeOffset End { get; init; }
    public string? ExperimentId { get; init; }
    public string? LayerId { get; init; }
    public int TrafficPercent { get; init; } = 100;
    public int TrafficOffset { get; init; } = 0;
    public string? AudienceFilters { get; init; }        // JSON: AudienceFilterEntry[]
    public string Method { get; init; } = "bayesian_ab";
}

/// <summary>
/// HTTP response body for POST /api/query/experiment.
/// Flat structure that maps variant names to per-variant statistics.
/// </summary>
public sealed class ExperimentQueryResponse
{
    [JsonPropertyName("metricType")]
    public required string MetricType { get; init; }

    [JsonPropertyName("variants")]
    public required Dictionary<string, VariantStatsDto> Variants { get; init; }
}

/// <summary>
/// Per-variant statistics. Binary metrics populate N and K; continuous metrics
/// populate N, Mean, Variance, and Total.
/// </summary>
public sealed class VariantStatsDto
{
    [JsonPropertyName("n")]
    public long N { get; init; }

    [JsonPropertyName("k")]
    public long? K { get; init; }           // binary only

    [JsonPropertyName("mean")]
    public double? Mean { get; init; }       // continuous only

    [JsonPropertyName("variance")]
    public double? Variance { get; init; }   // continuous only

    [JsonPropertyName("total")]
    public double? Total { get; init; }      // continuous only
}

File: agent/tsdb-cloudflare/src/models/dtos.ts
// ── SDK track payload (matches .NET TrackPayload) ─────────────────────────────

export interface TrackPayload {
  user: EndUserDto;
  variations: FlagEvalDto[];
  metrics: MetricEventDto[];
}

export interface EndUserDto {
  keyId: string;
  name?: string;
  properties?: Record<string, string>;
}

export interface FlagEvalDto {
  flagKey: string;
  variant: string;
  sendToExperiment?: boolean;
  experimentId?: string;
  layerId?: string;
  timestamp: number; // unix seconds
}

export interface MetricEventDto {
  eventName: string;
  numericValue?: number | null;
  type?: string;
  route?: string;
  appType?: string;
  timestamp: number; // unix seconds
  props?: Record<string, unknown>;
}

// ── Query request / response DTOs ─────────────────────────────────────────────

export interface ExperimentQueryRequest {
  envId: string;
  flagKey: string;
  metricEvent: string;
  metricType: string; // binary | revenue | count | duration
  metricAgg?: string; // once | sum | mean | count | latest
  controlVariant: string;
  treatmentVariant: string;
  start: string; // ISO 8601
  end: string;   // ISO 8601
  experimentId?: string;
  layerId?: string;
  trafficPercent?: number;
  trafficOffset?: number;
  audienceFilters?: string; // JSON: AudienceFilter[]
  method?: string; // bayesian_ab | bandit
}

export interface ExperimentManyQueryRequest extends ExperimentQueryRequest {
  guardrailEvents?: string[];
}

export interface ExperimentQueryResponse {
  metricType: string;
  variants: Record<string, VariantStatsDto>;
}

export type ExperimentManyQueryResponse = Record<string, ExperimentQueryResponse>;

export interface VariantStatsDto {
  n: number;
  k?: number;        // binary only
  mean?: number;     // continuous only
  variance?: number; // continuous only
  total?: number;    // continuous only
}

// ── Exposure entry ────────────────────────────────────────────────────────────

export interface ExposureEntry {
  firstExposedAt: number; // unix ms
  variant: string;
}

// ── Audience filter ───────────────────────────────────────────────────────────

export interface AudienceFilter {
  property: string;
  op: "eq" | "neq" | "in" | "nin";
  value?: string;
  values?: string[];
}

export function audienceFilterMatches(
  filter: AudienceFilter,
  props: Record<string, string> | null,
): boolean {
  if (!props || !(filter.property in props)) {
    return filter.op === "neq" || filter.op === "nin";
  }
  const actual = props[filter.property];
  switch (filter.op) {
    case "eq":  return actual === filter.value;
    case "neq": return actual !== filter.value;
    case "in":  return filter.values?.includes(actual) === true;
    case "nin": return filter.values?.includes(actual) !== true;
    default:    return true;
  }
}

// ── Experiment query (internal) ───────────────────────────────────────────────

export interface ExperimentQuery {
  envId: string;
  flagKey: string;
  metricEvent: string;
  metricType: string;
  metricAgg: string;
  controlVariant: string;
  treatmentVariants: string[];
  allVariants: string[];
  startMs: number;
  endMs: number;
  startDate: string; // yyyy-MM-dd
  endDate: string;   // yyyy-MM-dd
  experimentId: string | null;
  layerId: string | null;
  trafficPercent: number;
  trafficOffset: number;
  audienceFilters: AudienceFilter[] | null;
  method: string;
}

export function buildExperimentQuery(req: ExperimentQueryRequest): ExperimentQuery {
  const start = new Date(req.start);
  const end = new Date(req.end);
  const treatments = [req.treatmentVariant];
  const audienceFilters = req.audienceFilters
    ? JSON.parse(req.audienceFilters) as AudienceFilter[]
    : null;

  return {
    envId: req.envId,
    flagKey: req.flagKey,
    metricEvent: req.metricEvent,
    metricType: req.metricType,
    metricAgg: req.metricAgg ?? "once",
    controlVariant: req.controlVariant,
    treatmentVariants: treatments,
    allVariants: [req.controlVariant, ...treatments],
    startMs: start.getTime(),
    endMs: end.getTime(),
    startDate: toDateString(start),
    endDate: toDateString(end),
    experimentId: req.experimentId ?? null,
    layerId: req.layerId ?? null,
    trafficPercent: req.trafficPercent ?? 100,
    trafficOffset: req.trafficOffset ?? 0,
    audienceFilters,
    method: req.method ?? "bayesian_ab",
  };
}

function toDateString(d: Date): string {
  return d.toISOString().slice(0, 10);
}
